src/Adapter/Carrier/Repository/CarrierRangeRepository.php

```diff
@@ -0,0 +1,213 @@
+
+/**
+ * Provides access to carrier range data source
+ */
+class CarrierRangeRepository
+{
+    public function __construct(
+        protected readonly Connection $connection,
+        protected readonly string $dbPrefix,
+        protected readonly CarrierRepository $carrierRepository,
+    ) {
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     *
+     * @return array<int, array<int|string>>
+     *
+     * @throws CarrierException
+     * @throws \Doctrine\DBAL\Exception
+     * @throws \PrestaShop\PrestaShop\Core\Domain\AttributeGroup\Attribute\Exception\AttributeNotFoundException
+     * @throws CarrierConstraintException
+     * @throws \PrestaShop\PrestaShop\Core\Domain\Zone\Exception\ZoneException
+     * @throws \PrestaShop\PrestaShop\Core\Exception\CoreException
+     */
+    public function get(CarrierId $carrierId, ShopConstraint $shopConstraint): array
+    {
+        // Check shop constraint
+        $this->assertShopConstraint($shopConstraint);
+
+        // Get carrier
+        $carrier = $this->carrierRepository->get($carrierId);
+
+        // Then, create query to retrieve carrier ranges
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('cd.id_zone, cr.delimiter1 AS range_from, cr.delimiter2 AS range_to, cd.price AS range_price')
+            ->from($this->dbPrefix . 'delivery', 'cd')
+            ->andWhere('cd.id_carrier = :carrierId')
+            ->setParameter('carrierId', $carrierId->getValue());
+
+        // Apply shipping method (weight or price) and shop constraint, then execute the query
+        $this->applyRangeTypeForQuery($qb, $carrier);
+        $this->applyShopConstraint($qb, $shopConstraint);
+        $results = $qb->executeQuery()->fetchAllAssociative();
+
+        return $results;
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     * @param CarrierRangesCollectionInterface $rangesCollection
+     * @param ShopConstraint $shopConstraint
+     */
+    public function set(CarrierId $carrierId, CarrierRangesCollectionInterface $rangesCollection, ShopConstraint $shopConstraint): void
+    {
+        // Check shop constraint
+        $this->assertShopConstraint($shopConstraint);
+
+        // Get carrier
+        $rangeTable = $this->getRangeMethodTable($rangesCollection->getCalculatingMethod());
+
+        // Use transaction to ensure data consistency
+        $this->connection->beginTransaction();
+
+        // Reset carrier ranges
+        $this->reset($carrierId, $shopConstraint);
+
+        // Foreach zones
+        $rangesIds = [];
+        foreach ($rangesCollection->getZones() as $zone) {
+            $idZone = $zone->getZoneId();
+
+            // Foreach ranges in current zone
+            foreach ($zone->getRanges() as $range) {
+                // Check if range already exist in database
+                if (!in_array($range->getFrom() . '-' . $range->getTo(), array_keys($rangesIds), true)) {
```

Comment:
```suggestion
                // Check if range has already been added in database
                $rangeKey = $range->getFrom() . '-' . $range->getTo();
                if (!in_array($rangeKey, array_keys($rangesIds), true)) {
```

Reply:
Suggestion to understand this is used as a key and to only generate this key once

Also, I was wondering doesn't the range depend on the zone as well? Each zones have their own range even when they are equal no? So maybe the key should be:
```php
$rangeKey = $range->getFrom() . '-' . $range->getTo() . '-' . $idZone;
```

And although we don't handle it for now, later it should also include the shop constraint no?

Reply:
To document